Add smite and award reputation dialogs

ArmandoGraterol opened the pull request:
![FireShot Capture 799 - Colony CDapp - localhost](https://user-images.githubusercontent.com/18473896/214445149-db0e2fa4-cf52-4f2b-9345-a4a3bee96c06.png)
![FireShot Capture 800 - Colony CDapp - localhost](https://user-images.githubusercontent.com/18473896/214445154-6ab8584b-2a25-4233-a19e-c5e5f1467e5a.png)
![FireShot Capture 801 - Colony CDapp - localhost](https://user-images.githubusercontent.com/18473896/214445159-90547f80-1608-4e7a-92d1-f55aef3e4c38.png)

Contributes to #58

rdig commented on src/components/common/Dialogs/AwardAndSmiteDialogs/ManageReputationContainer/ManageReputationContainer.tsx:
```diff
@@ -0,0 +1,190 @@
+  ManageReputationDialogFormValues,
+} from '../types';
+
+const displayName = 'common.ColonyHome.ManageReputationContainer';
```
This should be in `Dialogs` rather than `ColonyHome`

rdig commented on src/components/common/Dialogs/AwardAndSmiteDialogs/ManageReputationDialogForm/TeamDropdownItem/TeamDropdownItem.tsx:
```diff
@@ -0,0 +1,49 @@
+  totalReputation: string | undefined;
+}
+
+const displayName = `common.ColonyHome.SmiteDialog.TeamDropdownItem`;
```
Should be in `Dialogs`

rdig commented on src/components/common/Dialogs/AwardAndSmiteDialogs/SmiteDialog/SmiteDialog.tsx:
```diff
@@ -0,0 +1,14 @@
+import { AwardAndSmiteDialogProps } from '../types';
+import ManageReputationContainer from '../ManageReputationContainer';
+
+const displayName = 'common.ColonyHome.SmiteDialog';
```
`Dialogs` again

rdig reviewed (approved):
Yep, all good here _(other than the names of the components)_

jakubcolony commented on src/components/common/Dialogs/AwardAndSmiteDialogs/ManageReputationDialogForm/ManageReputationDialogForm.tsx:
```diff
@@ -0,0 +1,329 @@
+// import NotEnoughReputation from '~dashboard/NotEnoughReputation';
+import { REPUTATION_LEARN_MORE } from '~constants/externalUrls';
+
+import { ColonyWatcher, User } from '~types/index';
```
Probably don't need to include `index`

jakubcolony commented on src/components/common/Dialogs/AwardAndSmiteDialogs/ManageReputationDialogForm/ReputationAmountInput/ReputationAmountInput.tsx:
```diff
@@ -0,0 +1,122 @@
+
+import { HookFormInput as Input } from '~shared/Fields';
+import Numeral from '~shared/Numeral';
+import { Colony } from '~types/index';
```
Probably don't need to include `index`

jakubcolony commented on src/components/shared/NoPermissionMessage/NoPermissionMessage.tsx:
```diff
@@ -22,7 +22,7 @@ const MSG = defineMessages({
 interface Props {
   requiredPermissions: ColonyRole[];
-  domainName?: string;
+  domainName?: string | null;
```
Can this ever be null? If `domain.metadata` is present, the name will be there
<img width="337" alt="image" src="https://user-images.githubusercontent.com/112586815/224106811-52f80ae4-13f7-4de7-8cab-bf63cede0242.png">

ArmandoGraterol replied:
It can be, in this case inside `TransferFundsDialogForm`:
```
          <NoPermissionMessage
            requiredPermissions={[ColonyRole.Funding]}
            domainName={
              hasRoleInFromDomain
                ? fromDomain?.metadata?.name ||
                  `Domain #${fromDomain?.nativeId}`
                : toDomain?.metadata?.name || `Domain #${toDomain?.nativeId}`
            }
          />
```

`fromDomain` and `toDomain` come from here:
```
  const fromDomain = colonyDomains.find(
    (domain) => domain?.nativeId === fromDomainId,
  );
```

```
  const toDomain = colonyDomains.find(
    (domain) => domain?.nativeId === toDomainId,
  );
```

So unfortunately even if `.metadata` wasn't optional, the `.find` assumes that it could be `null` in the case that it doesn't find the domain.

jakubcolony replied:
I don't think that's the case (although I might be wrong). If `.find` doesn't find anything, it'll return `undefined` instead of `null`. Removing `null` from the prop type doesn't trigger any type errors

ArmandoGraterol replied:
You are right, it does return `undefined`. I do remember adding the `null` because at some point Typescript was yelling me about it :thinking: But that doesn't seem to be case for `TransferFundsDialogForm` anymore

jakubcolony reviewed (changes requested):
Good start @ArmandoGraterol but I spotted a couple of issues with the dialog:

- There is something off with the reputation in the domain selector:

https://user-images.githubusercontent.com/112586815/224078322-6ff61feb-63f3-41c9-a0b8-f4b22ea1721d.mov

- Clicking on the Max button causes an error in the console:
<img width="468" alt="image" src="https://user-images.githubusercontent.com/112586815/224086599-2ac19153-9037-423e-ba10-fd6b8c365eda.png">

- For some reason, there seems to be an upper limit of 999 reputation points (above which the button becomes inactive). I don't know if it's intentional but I can put a higher value in the Dapp without it complaining

willm30 commented:
> * Clicking on the Max button causes an error in the console:

@Julianahlk  experienced the same in #288. The [advice here](https://github.com/nosir/cleave.js/issues/601) suggests downgrading to 1.4.10, which I believe worked (Juliana pls confirm 🙂)

ArmandoGraterol commented:
@jakubcolony The changes you suggested should be up! There are still some that will be added that are on the "Manage teams" PR, but I'm pinging you anyways in case you want to check that the current code is good with you.

@willm30 @Julianahlk That was also the same issue here, I had to downgrade `cleave.js`. Thank you all!

jakubcolony commented on src/components/common/Dialogs/AwardAndSmiteDialogs/ManageReputationContainer/ManageReputationContainer.tsx:
```diff
@@ -0,0 +1,186 @@
+    setUserReputation(userRepPercentage);
+  };
+
+  const { isVotingReputationEnabled } = useEnabledExtensions(colony);
```
I thought you'd pass this as a prop instead of making `useEnabledExtensions` accept `colony` as its argument

ArmandoGraterol replied:
Yeah, this branch doesn't have those changes. Until now that I'm rebasing it

jakubcolony reviewed (changes requested):
Hey @ArmandoGraterol, I had another look and spotted a couple of issues:

The domain selector still seems to be buggy (showing all the domains' reputation as the currently selected domain's reputation):
https://user-images.githubusercontent.com/112586815/226297854-5c65f000-ab26-42e4-9956-4f82a40420b3.mov

I can award negative reputation, is that intentional?
<img width="492" alt="image" src="https://user-images.githubusercontent.com/112586815/226298320-388c4c9c-68f0-43f4-93fe-0263fa4c9fd2.png">

I see some errors in the console when the input value is `.` or `-` but that will presumably go away when my fix for Cleave using raw value is merged - we'll be able to change the validation from using a custom test and `Decimal.js` to native yup methods

There's a problem with the smite reputation Max button, it seems like the validation doesn't re-run when it's clicked:

https://user-images.githubusercontent.com/112586815/226301188-9643b804-e377-4343-bc56-7819b3557cab.mov

ArmandoGraterol commented:
@jakubcolony Hey! I think I've addressed all of the issues you found except for the error when you input `.`, in case the fix you uploaded in the other PR deals with that.

Feel free to re-review and let me know if the fixes are working for you

jakubcolony commented on src/components/common/Dialogs/AwardAndSmiteDialogs/ManageReputationContainer/ManageReputationContainer.tsx:
```diff
@@ -0,0 +1,187 @@
+    );
+  }
+
+  const nativeTokenDecimals = nativeToken?.decimals || DEFAULT_TOKEN_DECIMALS;
```
Could you use the `getTokenDecimalsWithFallback` method instead?

ArmandoGraterol replied:
True, I forgot about this function

jakubcolony commented on src/components/common/Dialogs/AwardAndSmiteDialogs/ManageReputationContainer/ManageReputationContainer.tsx:
```diff
@@ -0,0 +1,187 @@
+      transform={transform}
+    >
+      {({ getValues }) => {
+        const values = getValues();
```
Should that be `watch`?

jakubcolony reviewed (approved):
Nice work!

There's just a few `getValues` that you could probably change to `watch` but other issues have been fixed

>EDIT: Forgot to mention, also I haven't been able to fix the issue with the smite reputation max button, as for now, it seems like userReputation is not being updated fast enough and therefore when you select a new user or domain, the user reputation used by the schema is the old one. Still looking into a solution for this one, figured I let you know in case you have any ideas or suggestions to fix this.

Could you possibly move the logic for getting the user reputation to the schema test? Also I think you'll need to call `trigger` whenever the selected user/team changes to re-run the validation
